File: Notchly/Panel/NotchShape.swift
import SwiftUI

struct NotchShape: Shape, Animatable {
    var edge: ScreenEdge
    /// Rounding on the two corners furthest from the screen edge.
    var cornerRadius: CGFloat
    /// Radius of the concave flare where the panel meets the edge.
    var inverseRadius: CGFloat

    init(edge: ScreenEdge, cornerRadius: CGFloat, inverseRadius: CGFloat) {
        self.edge = edge
        self.cornerRadius = cornerRadius
        self.inverseRadius = inverseRadius
    }

    var animatableData: AnimatablePair<CGFloat, CGFloat> {
        get { AnimatablePair(cornerRadius, inverseRadius) }
        set { cornerRadius = newValue.first; inverseRadius = newValue.second }
    }

    func path(in rect: CGRect) -> Path {
        // Canonical space is always "hanging from the top"; for the side edges the
        // canonical rect is the transposed one.
        let canonicalSize = edge.growsHorizontally
            ? CGSize(width: rect.height, height: rect.width)
            : rect.size
        var path = Self.canonicalPath(size: canonicalSize,
                                      cornerRadius: cornerRadius,
                                      inverseRadius: inverseRadius)
        path = path.applying(Self.transform(for: edge, in: rect.size))
        return path.offsetBy(dx: rect.minX, dy: rect.minY)
    }

    /// A top-docked notch: concave flares at `y == 0`, rounded corners at the bottom.
    private static func canonicalPath(size: CGSize, cornerRadius: CGFloat, inverseRadius: CGFloat) -> Path {
        let w = size.width
        let h = size.height
        // Clamp so extreme settings (or mid-animation values) can never self-intersect.
        let ir = max(0, min(inverseRadius, w / 2, h))
        let cr = max(0, min(cornerRadius, (w - 2 * ir) / 2, h - ir))

        var path = Path()
        guard w > 0, h > 0 else { return path }

        path.move(to: CGPoint(x: 0, y: 0))
        // Flare in from the edge to the left wall of the body.
        path.addQuadCurve(to: CGPoint(x: ir, y: ir), control: CGPoint(x: ir, y: 0))
        path.addLine(to: CGPoint(x: ir, y: h - cr))
        path.addQuadCurve(to: CGPoint(x: ir + cr, y: h), control: CGPoint(x: ir, y: h))
        path.addLine(to: CGPoint(x: w - ir - cr, y: h))
        path.addQuadCurve(to: CGPoint(x: w - ir, y: h - cr), control: CGPoint(x: w - ir, y: h))
        path.addLine(to: CGPoint(x: w - ir, y: ir))
        // Flare back out to the edge.
        path.addQuadCurve(to: CGPoint(x: w, y: 0), control: CGPoint(x: w - ir, y: 0))
        path.closeSubpath()
        return path
    }

    /// Maps canonical (top-docked) coordinates onto the requested edge.
    private static func transform(for edge: ScreenEdge, in size: CGSize) -> CGAffineTransform {
        switch edge {
        case .top:
            return .identity
        case .bottom:
            return CGAffineTransform(a: 1, b: 0, c: 0, d: -1, tx: 0, ty: size.height)
        case .trailing:
            return CGAffineTransform(a: 0, b: 1, c: -1, d: 0, tx: size.width, ty: 0)
        case .leading:
            return CGAffineTransform(a: 0, b: 1, c: 1, d: 0, tx: 0, ty: 0)
        }
    }
}

File: ui/lib/notch-shape.js
// The Panel outline, ported from the Swift `NotchShape`.
//
// Two of its corners are concave: they flare outward where the panel meets the screen
// edge, which is what makes the shape read as part of the bezel instead of a window
// floating on top of it. The geometry is authored once for a top-docked panel and then
// transformed for the other three edges, so all four sides stay identical.

/** @typedef {"top" | "bottom" | "leading" | "trailing"} ScreenEdge */

/** True on the left and right edges, where the panel extends horizontally. */
export function growsHorizontally(edge) {
  return edge === "leading" || edge === "trailing";
}

const clamp = (value, min, max) => Math.max(min, Math.min(value, max));

/**
 * Builds the canonical top-docked outline: concave flares at y = 0, rounded corners at
 * the bottom. Returned as a list of path commands in canonical space.
 */
function canonicalPath(width, height, cornerRadius, inverseRadius) {
  const ir = clamp(inverseRadius, 0, Math.min(width / 2, height));
  const cr = clamp(cornerRadius, 0, Math.min((width - 2 * ir) / 2, height - ir));

  return [
    ["M", 0, 0],
    ["Q", ir, 0, ir, ir],
    ["L", ir, height - cr],
    ["Q", ir, height, ir + cr, height],
    ["L", width - ir - cr, height],
    ["Q", width - ir, height, width - ir, height - cr],
    ["L", width - ir, ir],
    ["Q", width - ir, 0, width, 0],
    ["Z"],
  ];
}

/** Maps canonical (top-docked) coordinates onto the requested edge. */
function transformFor(edge, width, height) {
  switch (edge) {
    case "top":
      return (x, y) => [x, y];
    case "bottom":
      return (x, y) => [x, height - y];
    case "trailing":
      return (x, y) => [width - y, x];
    case "leading":
      return (x, y) => [y, x];
    default:
      throw new Error(`unknown edge: ${edge}`);
  }
}

/**
 * SVG path data for the panel outline.
 *
 * @param {ScreenEdge} edge which side of the display the panel docks to
 * @param {number} width  drawing width, including the flares
 * @param {number} height drawing height, including the flares
 */
export function notchPath(edge, width, height, cornerRadius, inverseRadius) {
  if (width <= 0 || height <= 0) return "";

  // On the side edges the canonical rect is the transposed one.
  const [cw, ch] = growsHorizontally(edge) ? [height, width] : [width, height];
  const commands = canonicalPath(cw, ch, cornerRadius, inverseRadius);
  const map = transformFor(edge, width, height);
  const round = (n) => Math.round(n * 1000) / 1000;

  return commands
    .map((command) => {
      const [op, ...values] = command;
      if (op === "Z") return "Z";
      const points = [];
      for (let i = 0; i < values.length; i += 2) {
        const [x, y] = map(values[i], values[i + 1]);
        points.push(`${round(x)} ${round(y)}`);
      }
      return `${op}${points.join(" ")}`;
    })
    .join("");
}
